### PR TITLE
Rewards vesting

### LINKED ISSUE #9007: Implement rewards vesting
Build the vesting engine + snapshot.

See pending PR: https://github.com/vegaprotocol/specs/pull/1876

### PR DESCRIPTION
**Pull request by jeremyletang**
close #9007

### CODE REVIEW

**Review by ze97286 — COMMENTED** [review]
does it not require checkpoint support?

File: CHANGELOG.md [hunk 1]
```diff
@@ -48,6 +48,7 @@
 - [8763](https://github.com/vegaprotocol/vega/issues/8763) - Periodic settlement data endpoint.
 - [8920](https://github.com/vegaprotocol/vega/issues/8920) - Emit events when something happens to teams.
 - [8917](https://github.com/vegaprotocol/vega/issues/8917) - Support teams engine snapshots
+- [9007](https://github.com/vegaprotocol/vega/issues/9007) - Add reward vesting mecanisms.
```
Comment (candida-d) [inline CHANGELOG.md:51]:
```suggestion
- [9007](https://github.com/vegaprotocol/vega/issues/9007) - Add reward vesting mechanisms.
```

File: protos/sources/vega/vega.proto [hunk 2]
```diff
@@ -792,6 +792,12 @@ enum AccountType {
   // Network controlled treasury
   ACCOUNT_TYPE_NETWORK_TREASURY = 21;
 
+  // An account holding users reward for the vesting period
+  ACCOUNT_TYPE_VESTING_REWARDS = 22;
+
+  // An account holding users reward after the vesting period
```
Comment (candida-d) [inline protos/sources/vega/vega.proto:798]:
```suggestion
  // Account holding user's rewards for the vesting period
  ACCOUNT_TYPE_VESTING_REWARDS = 22;

  // Account holding user's rewards after the vesting period
```

File: core/collateral/engine.go [hunk 3]
```diff
@@ -3113,6 +3157,62 @@ func (e *Engine) CreatePartyGeneralAccount(ctx context.Context, partyID, asset s
 	return generalID, nil
 }
 
+// GetOrCreatePartyVestingAccount create the general account for a party.
+func (e *Engine) GetOrCreatePartyVestingRewardAccount(ctx context.Context, partyID, asset string) *types.Account {
+	if !e.AssetExists(asset) {
+		e.log.Panic("trying to use an unexisting asset for reward accounts, something went very wrong somewhere",
```
Comment (candida-d) [inline core/collateral/engine.go:3163]:
```suggestion
		e.log.Panic("trying to use a nonexisting asset for reward accounts, something went very wrong somewhere",
```

File: core/collateral/engine.go [hunk 4]
```diff
@@ -3113,6 +3157,62 @@ func (e *Engine) CreatePartyGeneralAccount(ctx context.Context, partyID, asset s
...
+		e.broker.Send(events.NewAccountEvent(ctx, acc))
+	}
+
+	return acc
+}
+
+// GetOrCreatePartyVestedAccount create the general account for a party.
+func (e *Engine) GetOrCreatePartyVestedRewardAccount(ctx context.Context, partyID, asset string) *types.Account {
+	if !e.AssetExists(asset) {
+		e.log.Panic("trying to use an unexisting asset for reward accounts, something went very wrong somewhere",
```
Comment (candida-d) [inline core/collateral/engine.go:3191]:
```suggestion
		e.log.Panic("trying to use a nonexisting asset for reward accounts, something went very wrong somewhere",
```

File: core/collateral/checkpoint.go [hunk 5]
```diff
@@ -106,14 +113,31 @@ func (e *Engine) Load(ctx context.Context, data []byte) error {
...
-		lm, err := e.RestoreCheckpointBalance(
-			ctx, noMarket, balance.Party, balance.Asset, types.AccountTypeGeneral, ub.Clone())
-		if err != nil {
-			return err
+		var (
+			lm  *types.LedgerMovement
+			err error
+		)
+
+		if isVesting {
```
Comment (ze97286) [inline core/collateral/checkpoint.go:121]: can you add a test for this logic please

File: core/protocol/all_services.go [hunk 6]
```diff
@@ -335,7 +339,7 @@ func newServices(
 
 	svcs.snapshotEngine.AddProviders(svcs.checkpoint, svcs.collateral, svcs.governance, svcs.delegation, svcs.netParams, svcs.epochService, svcs.assets, svcs.banking, svcs.witness,
 		svcs.notary, svcs.stakingAccounts, svcs.stakeVerifier, svcs.limits, svcs.topology, svcs.eventForwarder, svcs.executionEngine, svcs.marketActivityTracker, svcs.statevar,
-		svcs.erc20MultiSigTopology, svcs.protocolUpgradeEngine, svcs.ethereumOraclesVerifier)
+		svcs.erc20MultiSigTopology, svcs.protocolUpgradeEngine, svcs.ethereumOraclesVerifier, svcs.vesting)
```
Comment (ze97286) [inline core/protocol/all_services.go:342]: you didn't wire it to the checkpoint in line 304

Comment (jeremyletang) [reply core/protocol/all_services.go:342]: yea nothing is restored from the checkpoint, only the balances in the collateral.

Comment (ze97286) [reply core/protocol/all_services.go:342]: ah, cool

File: core/collateral/checkpoint.go (hunk 5 above)
Comment (jeremyletang) [reply core/collateral/checkpoint.go:121]: good point, and just at looking at the code I see a bug...

Comment (ze97286) [reply core/collateral/checkpoint.go:121]: that was my way of telling you there was one ;)

Comment (jeremyletang) [reply core/collateral/checkpoint.go:121]: addressed

File: core/vesting/vesting.go [hunk 7]
```diff
@@ -0,0 +1,284 @@
...
+) *Engine {
+	log = log.Named(namedLogger)
+
+	return &Engine{
+		log:         log,
+		c:           c,
+		asvm:        asvm,
+		broker:      broker,
+		assets:      assets,
+		baseRate:    num.MustDecimalFromString("0.01"),
```
Comment (ze97286) [inline core/vesting/vesting.go:87]: why do you need this initialised here? it should never be used with this value anyways

File: core/protocol/all_services.go [hunk 8]
```diff
@@ -292,6 +294,8 @@ func newServices(
 		svcs.delegation = delegation.New(svcs.log, svcs.conf.Delegation, svcs.broker, svcs.topology, svcs.stakingAccounts, svcs.epochService, svcs.timeService)
 	}
 
+	svcs.vesting = vesting.NewSnapshotEngine(svcs.log, svcs.collateral, DummyASVM{}, svcs.broker, svcs.assets)
+	svcs.epochService.NotifyOnEpoch(svcs.topology.OnEpochEvent, svcs.vesting.OnEpochEvent)
```
[the remark this answers is not in the document]
Comment (jeremyletang) [reply core/protocol/all_services.go]: fixed

File: core/vesting/vesting.go (hunk 7 above)
Comment (jeremyletang) [reply core/vesting/vesting.go]: removed.

File: core/types/snapshot_nodes.go [hunk 9]
```diff
@@ -1069,6 +1071,7 @@ func (p Payload) IntoProto() *snapshot.Payload {
 	case *snapshot.Payload_Teams:
 		ret.Data = dt
 	case *snapshot.Payload_TeamSwitches:
+	case *snapshot.Payload_Vesting:
 		ret.Data = dt
```
Comment (wwestgarth) [inline core/types/snapshot_nodes.go:1075]: Think we're now missing the `ret.Data = dt` for `Payload_TeamSwitches`.

Comment (jeremyletang) [reply core/types/snapshot_nodes.go:1075]: oh a terrible rebase again.

Comment (jeremyletang) [reply core/types/snapshot_nodes.go:1076]: addressed

File: core/vesting/vesting.go [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+
+func (e *Engine) increaseVestingBalance(
+	party, asset string,
+	amount *num.Uint,
+) {
+	partyRewards := e.getPartyRewards(party)
+
+	vesting, ok := partyRewards.Vesting[asset]
+	if !ok {
+		vesting = num.NewUint(0)
```
[the remark this answers is not in the document]
Comment (jeremyletang) [reply core/vesting/vesting.go]: addressed

File: core/protocol/all_services.go [hunk 11]
```diff
@@ -292,6 +294,8 @@ func newServices(
 		svcs.delegation = delegation.New(svcs.log, svcs.conf.Delegation, svcs.broker, svcs.topology, svcs.stakingAccounts, svcs.epochService, svcs.timeService)
 	}
 
+	svcs.vesting = vesting.NewSnapshotEngine(svcs.log, svcs.collateral, DummyASVM{}, svcs.broker, svcs.assets)
+	svcs.epochService.NotifyOnEpoch(svcs.vesting.OnEpochEvent, svcs.vesting.OnEpochRestore)
```
Comment (ze97286) [inline core/protocol/all_services.go:298]: I *think* this must happen after rewards is constructed to ensure that the reward engine is called first for onEpochEvent.

Comment (jeremyletang) [reply core/protocol/all_services.go:298]: right. makes sense.

File: core/collateral/engine.go [hunk 12]
```diff
@@ -685,6 +685,50 @@ func (e *Engine) TransferRewards(ctx context.Context, rewardAccountID string, tr
...
+) ([]*types.LedgerMovement, error) {
+	if len(transfers) == 0 {
+		return nil, nil
+	}
+
+	transferReqs := make([]*types.TransferRequest, 0, len(transfers))
+	for _, t := range transfers {
+		transferReqs = append(transferReqs, &types.TransferRequest{
+			FromAccount: []*types.Account{
+				e.GetOrCreatePartyVestingRewardAccount(ctx, t.Owner, t.Amount.Asset),
```
Comment (ze97286) [inline core/collateral/engine.go:699]: is it possible for it not to exist?

Comment (jeremyletang) [reply core/collateral/engine.go:699]: nah, that's why I make it panic internally, so I thought i'd be good enough.

**Review by ze97286 — APPROVED** [review]

**Pull request merged**